File: baslercam/controller_cpp/src/basler_driver.cpp
#include "basler_ros2_driver/basler_driver.hpp"

namespace basler
{
    BaslerROS2Driver::BaslerROS2Driver(const rclcpp::NodeOptions &options) : Node("basler_ros2_driver", options)
    {
        RCLCPP_INFO(this->get_logger(), "started basler_ros2_driver Node");
        _open_basler_cameras_server = this->create_service<Trigger>("open_basler_cameras", std::bind(&BaslerROS2Driver::_openBaslerCamerasCb, this, std::placeholders::_1, std::placeholders::_2));
        _close_basler_cameras_server = this->create_service<Trigger>("close_basler_cameras", std::bind(&BaslerROS2Driver::_closeBaslerCamerasCb, this, std::placeholders::_1, std::placeholders::_2));
        _get_color_image_server = this->create_service<GetColorImage>("get_basler_color_image", std::bind(&BaslerROS2Driver::_getColorImageCb, this, std::placeholders::_1, std::placeholders::_2));
        _get_mono_images_server = this->create_service<GetMonoImages>("get_basler_mono_images", std::bind(&BaslerROS2Driver::_getMonoImagesCb, this, std::placeholders::_1, std::placeholders::_2));
        _get_all_images_server = this->create_service<GetAllImages>("get_basler_all_images", std::bind(&BaslerROS2Driver::_getAllImagesCb, this, std::placeholders::_1, std::placeholders::_2));
        _bgr_color_publisher = this->create_publisher<Image>("/basler/color/image_raw", rclcpp::SensorDataQoS());
        Pylon::PylonInitialize();
    }
    BaslerROS2Driver::~BaslerROS2Driver()
    {

        if (_avena_basler_cameras != nullptr && (_avena_basler_cameras->IsOpen() || _avena_basler_cameras->IsGrabbing()))
        {
            _closeBaslerCameras();
        }
        rclcpp::shutdown();
    }

    void BaslerROS2Driver::_openBaslerCameras(const CamName_CamSerial &camera_group_serials)
    {
        RCLCPP_INFO(this->get_logger(), "Trying to open cameras");
        try
        {
            Pylon::CTlFactory &tlFactory = Pylon::CTlFactory::GetInstance();
            Pylon::DeviceInfoList_t devices;
            auto no_of_connected_devices = tlFactory.EnumerateDevices(devices);
            if (no_of_connected_devices == 0)
            {
                RCLCPP_ERROR(this->get_logger(), "can not access any cameras");
                return;
            }
            else if (no_of_connected_devices < _request_no_of_cameras)
            {
                RCLCPP_ERROR(this->get_logger(), "can not access all requested cameras");
                return;
            }
            else
            {
                RCLCPP_INFO_STREAM(this->get_logger(), "Number of connected cameras: " << no_of_connected_devices);
                _avena_basler_cameras = std::make_shared<Pylon::CInstantCameraArray>();
                _avena_basler_cameras->Initialize(devices.size());
                for (size_t camera_idx = 0; camera_idx < _avena_basler_cameras->GetSize(); camera_idx++)
                {
                    (*_avena_basler_cameras)[camera_idx].Attach(tlFactory.CreateDevice(devices[camera_idx]));
                    auto device_serial = std::string((*_avena_basler_cameras)[camera_idx].GetDeviceInfo().GetSerialNumber());
                    if (camera_group_serials.find(device_serial) != camera_group_serials.end())
                    {
                        RCLCPP_INFO_STREAM(this->get_logger(), camera_group_serials.at(device_serial) << " camera serial number: " << device_serial);
                        if (camera_group_serials.at(device_serial) == "color")
                        {
                            _color_handler = new ColorImageEventHandler(_bgr_color_publisher);
                            _color_config_handler = new ColorCameraConfigurationHandler();
                            (*_avena_basler_cameras)[camera_idx].RegisterImageEventHandler(_color_handler, Pylon::RegistrationMode_Append, Pylon::Cleanup_Delete);
                            (*_avena_basler_cameras)[camera_idx].RegisterConfiguration(_color_config_handler, Pylon::RegistrationMode_Append, Pylon::Cleanup_Delete);
                        }
                        else if (camera_group_serials.at(device_serial) == "left_mono")
                        {
                            _left_mono_handler = new LeftMonoImageEventHandler();
                            _left_mono_config_handler = new MonoCameraConfigurationHandler();
                            (*_avena_basler_cameras)[camera_idx].RegisterImageEventHandler(_left_mono_handler, Pylon::RegistrationMode_Append, Pylon::Cleanup_Delete);
                            (*_avena_basler_cameras)[camera_idx].RegisterConfiguration(_left_mono_config_handler, Pylon::RegistrationMode_Append, Pylon::Cleanup_Delete);
                        }
                        else if (camera_group_serials.at(device_serial) == "right_mono")
                        {
                            _right_mono_handler = new RightMonoEventHandler();
                            _right_mono_config_handler = new MonoCameraConfigurationHandler();
                            (*_avena_basler_cameras)[camera_idx].RegisterImageEventHandler(_right_mono_handler, Pylon::RegistrationMode_Append, Pylon::Cleanup_Delete);
                            (*_avena_basler_cameras)[camera_idx].RegisterConfiguration(_right_mono_config_handler, Pylon::RegistrationMode_Append, Pylon::Cleanup_Delete);
                        }
                    }
                    else
                    {
                        RCLCPP_WARN(this->get_logger(), "Unsupported Device connected");
                    }
                }
                if (_avena_basler_cameras->IsPylonDeviceAttached())
                {
                    _avena_basler_cameras->StartGrabbing(Pylon::GrabStrategy_LatestImageOnly, Pylon::GrabLoop_ProvidedByInstantCamera);
                    RCLCPP_INFO(this->get_logger(), "cameras are open");
                }
                else
                {
                    RCLCPP_WARN(this->get_logger(), "cameras are not open");
                    return ;
                }
                
            }
        }
        catch (const GenICam_3_1_Basler_pylon::GenericException &e)
        {
            RCLCPP_ERROR(this->get_logger(), e.what());
            return;
        }
    }

    void BaslerROS2Driver::_openBaslerCamerasCb(const std::shared_ptr<Trigger::Request> /*request*/,
                                                std::shared_ptr<Trigger::Response> response)
    {
        RCLCPP_INFO(this->get_logger(), "Open Basler cameras is called");
        if (_avena_basler_cameras == nullptr || _avena_basler_cameras->IsOpen() != true)
        {
            _openBaslerCameras(_camera_group_serials);
            if (_avena_basler_cameras != nullptr && _avena_basler_cameras->IsGrabbing())
            {
                response->success = true;
            }
            else
            {
                response->success = false;
            }
        }
        else
        {
            RCLCPP_WARN(this->get_logger(), "cameras are already opened");
            response->success = true;
        }
        if (_avena_basler_cameras != nullptr && _avena_basler_cameras->GetSize() == _camera_group_serials.size())
        {
            response->success = true;
        }
        else
        {
            response->success = false;
        }
    }

    void BaslerROS2Driver::_closeBaslerCameras()
    {
        if (_avena_basler_cameras->IsGrabbing())
        {
            _avena_basler_cameras->StopGrabbing();
        }
        else if (_avena_basler_cameras->IsOpen())
        {
            _avena_basler_cameras->Close();
        }

        for (size_t camera_idx = 0; camera_idx < _avena_basler_cameras->GetSize(); camera_idx++)
        {

            auto device_serial = std::string((*_avena_basler_cameras)[camera_idx].GetDeviceInfo().GetSerialNumber());
            if (_camera_group_serials.find(device_serial) != _camera_group_serials.end())
            {

                if (_camera_group_serials.at(device_serial) == "color")
                {

                    if (_color_handler != nullptr && _color_config_handler != nullptr)
                    {
                        (*_avena_basler_cameras)[camera_idx].DeregisterImageEventHandler(_color_handler);
                        (*_avena_basler_cameras)[camera_idx].DeregisterConfiguration(_color_config_handler);
                    }
                }
                else if (_camera_group_serials.at(device_serial) == "left_mono")
                {

                    if (_left_mono_handler != nullptr && _left_mono_config_handler != nullptr)
                    {

                        (*_avena_basler_cameras)[camera_idx].DeregisterImageEventHandler(_left_mono_handler);
                        (*_avena_basler_cameras)[camera_idx].DeregisterConfiguration(_left_mono_config_handler);
                    }
                }
                else if (_camera_group_serials.at(device_serial) == "right_mono")
                {

                    if (_right_mono_handler != nullptr && _right_mono_config_handler != nullptr)
                    {
                        (*_avena_basler_cameras)[camera_idx].DeregisterImageEventHandler(_right_mono_handler);
                        (*_avena_basler_cameras)[camera_idx].DeregisterConfiguration(_right_mono_config_handler);
                    }
                }
            }
        }

        RCLCPP_INFO(this->get_logger(), "cameras are closed");
    }
    void BaslerROS2Driver::_closeBaslerCamerasCb(const std::shared_ptr<Trigger::Request> /*request*/,
                                                 std::shared_ptr<Trigger::Response> response)
    {
        RCLCPP_INFO(this->get_logger(), "close Basler cameras is called");
        if (_avena_basler_cameras != nullptr && _avena_basler_cameras->GetSize() > 0 && _avena_basler_cameras->IsOpen())
        {
            _closeBaslerCameras();
            response->success = true;
        }
        else
        {
            RCLCPP_WARN(this->get_logger(), "cameras are already closed or not open");
            response->success = true;
        }
    }
    void BaslerROS2Driver::_getAllImagesCb(const std::shared_ptr<GetAllImages::Request> /*request*/,
                                           std::shared_ptr<GetAllImages::Response> response)
    {
        RCLCPP_INFO(this->get_logger(), "get all images is called");

        if (_left_mono_handler != nullptr && _right_mono_handler != nullptr && _color_handler != nullptr && _avena_basler_cameras != nullptr && _avena_basler_cameras->IsGrabbing())
        {
            auto left_mono_image = _left_mono_handler->getLeftMonoImage();
            auto right_mono_image = _right_mono_handler->getRightMonoimage();
            auto color_4k_image = _color_handler->get4kColorImage();

            if (left_mono_image.size().width > 0 && right_mono_image.size().width > 0 && color_4k_image.size().width > 0)
            {
                auto left_header = std_msgs::msg::Header();
                auto right_header = std_msgs::msg::Header();
                auto color_header = std_msgs::msg::Header();
                color_header.frame_id = "/basler/color/image_raw";
                left_header.frame_id = "/basler/left/image_raw";
                right_header.frame_id = "/basler/right/image_raw";
                auto now = builtin_interfaces::msg::Time(this->now());
                left_header.stamp = now;
                right_header.stamp = now;
                color_header.stamp = now;
                cv_bridge::CvImagePtr left_cv_ptr = std::make_shared<cv_bridge::CvImage>(left_header, sensor_msgs::image_encodings::MONO8, left_mono_image);
                cv_bridge::CvImagePtr right_cv_ptr = std::make_shared<cv_bridge::CvImage>(right_header, sensor_msgs::image_encodings::MONO8, right_mono_image);
                cv_bridge::CvImagePtr color_cv_ptr = std::make_shared<cv_bridge::CvImage>(color_header, sensor_msgs::image_encodings::RGB8, color_4k_image);

                try
                {
                    response->left_mono = *(left_cv_ptr->toImageMsg());
                    response->right_mono = *(right_cv_ptr->toImageMsg());
                    response->color = *(color_cv_ptr->toImageMsg());
                }
                catch (cv_bridge::Exception &e)
                {
                    RCLCPP_ERROR(this->get_logger(), e.what());
                    return;
                }
            }
        }
        else
        {
            RCLCPP_WARN(this->get_logger(), "cameras are not open, return empty response");
            response->left_mono = Image();
            response->right_mono = Image();
            response->color = Image();
        }
    }
    void BaslerROS2Driver::_getColorImageCb(const std::shared_ptr<GetColorImage::Request> /*request*/,
                                            std::shared_ptr<GetColorImage::Response> response)
    {
        RCLCPP_INFO(this->get_logger(), "get Color Image is called");
        if (_color_handler != nullptr && _avena_basler_cameras != nullptr && _avena_basler_cameras->IsGrabbing())
        {
            auto color_4k_image = _color_handler->get4kColorImage();
            if (color_4k_image.size().width > 0)
            {
                auto header = std_msgs::msg::Header();
                header.frame_id = "/basler/color/image_raw";
                header.stamp = builtin_interfaces::msg::Time(this->now());
                cv_bridge::CvImagePtr cv_ptr = std::make_shared<cv_bridge::CvImage>(header, sensor_msgs::image_encodings::RGB8, color_4k_image);
                try
                {
                    response->color = *(cv_ptr->toImageMsg());
                }
                catch (cv_bridge::Exception &e)
                {
                    RCLCPP_ERROR(this->get_logger(), e.what());
                    return;
                }
            }
        }
        else
        {
            RCLCPP_WARN(this->get_logger(), "cameras are not open, returning empty response");
            response->color = Image();
        }
    }
    void BaslerROS2Driver::_getMonoImagesCb(const std::shared_ptr<GetMonoImages::Request> /*request*/,
                                            std::shared_ptr<GetMonoImages::Response> response)
    {
        RCLCPP_INFO(this->get_logger(), "get Mono Images is called");
        if (_left_mono_handler != nullptr && _right_mono_handler != nullptr && _avena_basler_cameras != nullptr && _avena_basler_cameras->IsGrabbing())
        {
            auto left_mono_image = _left_mono_handler->getLeftMonoImage();
            auto right_mono_image = _right_mono_handler->getRightMonoimage();
            if (left_mono_image.size().width > 0 && right_mono_image.size().width > 0)
            {
                auto left_header = std_msgs::msg::Header();
                auto right_header = std_msgs::msg::Header();
                left_header.frame_id = "/basler/left/image_raw";
                right_header.frame_id = "/basler/right/image_raw";
                auto now = builtin_interfaces::msg::Time(this->now());
                left_header.stamp = now;
                right_header.stamp = now;
                cv_bridge::CvImagePtr left_cv_ptr = std::make_shared<cv_bridge::CvImage>(left_header, sensor_msgs::image_encodings::MONO8, left_mono_image);
                cv_bridge::CvImagePtr right_cv_ptr = std::make_shared<cv_bridge::CvImage>(right_header, sensor_msgs::image_encodings::MONO8, right_mono_image);
                try
                {
                    response->left_mono = *(left_cv_ptr->toImageMsg());
                    response->right_mono = *(right_cv_ptr->toImageMsg());
                }
                catch (cv_bridge::Exception &e)
                {
                    RCLCPP_ERROR(this->get_logger(), e.what());
                    return;
                }
            }
        }
        else
        {
            RCLCPP_WARN(this->get_logger(), "cameras are not open, returning empty response");
            response->left_mono = Image();
            response->right_mono = Image();
        }
    }
} // namespace basler
#include "rclcpp_components/register_node_macro.hpp"
RCLCPP_COMPONENTS_REGISTER_NODE(basler::BaslerROS2Driver)

File: baslercam/basler_driver.py
import cv2
from datetime import datetime

from pypylon import genicam
from pypylon import pylon

from custom_interfaces.srv import GetAllImages, GetMonoImages, GetColorImage

from std_srvs.srv import Trigger
from cv_bridge import CvBridge
from sensor_msgs.msg import Image
from rclpy.node import Node
import rclpy

class ColorImageEventHandler(pylon.ImageEventHandler):
    """color image event handler to manage grab result

    It is an event handler to register a callback for color camera grab result.

    Attributes:
        color_image: 4k bgr8 image
        color_hd_image: hd bgr8 image
    """
    def __init__(self, *args):
        super().__init__(*args)
        self.color_image = None
        self.color_hd_image = None

    def OnImageGrabbed(self, camera: pylon.InstantCamera, grab_result: pylon.GrabResult) -> None:
        """Image callback for color image
        
        It is activated each time, the camera grabs an image. 
        We rotate the image for 90 degrees counterclockwise for disparity estimation and resize 4k image to new hd image
        for publishing.
        
        Args:
            camera: instant basler camera object
            grab_result: basler camera grab result
        Returns:
            None
        Raises:
            CVError: An error occurred accessing the cv numpy array.
        """
        if grab_result.GrabSucceeded():
            temp = grab_result.GetArray()
            self.color_image = cv2.rotate(temp, cv2.ROTATE_90_COUNTERCLOCKWISE)
            self.color_hd_image = cv2.resize(self.color_image, (self.color_image.shape[1]//3, self.color_image.shape[0]//3), interpolation = cv2.INTER_AREA)
            grab_result.Release()

class LeftMonoImageEventHandler(pylon.ImageEventHandler):
    """left image event handler to manage grab result

    It is an event handler to register a callback for left mono camera grab result.

    Attributes:
        left_mono_image: fhd mono8 image
    """
    def __init__(self, *args):
        super().__init__(*args)
        self.left_mono_image = None

    def OnImageGrabbed(self, camera: pylon.InstantCamera, grab_result: pylon.GrabResult)->None:
        """Image callback for left mono image
        
        It is activated each time, the camera grabs an image. 
        We rotate the image for 90 degrees counterclockwise for disparity estimation.
        
        Args:
            camera: instant basler camera object
            grab_result: basler camera grab result
        Returns:
            None
        Raises:
            CVError: An error occurred accessing the cv numpy array
        """
        if grab_result.GrabSucceeded():
            temp = grab_result.GetArray()
            self.left_mono_image = cv2.rotate(temp, cv2.ROTATE_90_COUNTERCLOCKWISE)
            grab_result.Release()


class RightMonoEventHandler(pylon.ImageEventHandler):
    """right image event handler to manage grab result

    It is an event handler to register a callback for right mono camera grab result.

    Attributes:
        right_mono_image: fhd mono8 image 
    """
    def __init__(self, *args):
        super().__init__(*args)
        self.right_mono_image = None

    def OnImageGrabbed(self, camera: pylon.InstantCamera, grab_result: pylon.GrabResult)->None:
        """Image callback for mono image
        
        It is activated each time, the camera grabs an image. 
        We rotate the image for 90 degrees counterclockwise for disparity estimation
        
        Args:
            camera: instant basler camera object
            grab_result: basler camera grab result
        Returns:
            None
        Raises:
            CVError: An error occurred accessing the cv numpy array.
        """
        if grab_result.GrabSucceeded():
            temp = grab_result.GetArray()
            self.right_mono_image = cv2.rotate(temp, cv2.ROTATE_90_COUNTERCLOCKWISE)
            grab_result.Release()


class BaslerPyDriver(Node):
    """Basler ROS Node class for python

    It is an basler camera controller, providing services and topics for accessing cameras data

    Attributes:
        qos_profile: publisher qos profile
        camera_group_serials: user cameras serials 
        _camera_group_serials: saved user cameras serials
        _avena_basler_cameras: pypylon Instant cameras array for synchronization
        open_basler_cameras_service_server: ros service server to open cameras
        get_basler_color_image_srv: ros service server to get color image
        get_basler_mono_images_srv: ros service server to get two mono cameras
        get_basler_all_images_srv: ros service server to get 2 mono and 1 bgr images from basler camera
        basler_camera_color_pub: ros publisher to publish hd color image for security
        color_fps: timer frequency for publishing hd color image
        timer: timer ros object to publish image in timed callback
        bridge: converter of opencv to and from ros images
        _color_image_event_handler: image event handler for color camera
        _left_mono_image_event_handler: image event handler for left mono camera
        _right_mono_event_handler: image event handler for right mono camera

    Example:

        rclpy.init()
        camera_group_serials = {'left_mono': '40132642', 'color': '40120189', 'right_mono': '40134745'}
        basler_py_driver_node = BaslerPyDriver(qos_profile_sensor_data,camera_group_serials)
        try:
            while rclpy.ok():
                rclpy.spin_once(basler_py_driver_node)
                if basler_py_driver_node._avena_basler_cameras is not None and basler_py_driver_node._avena_basler_cameras.IsOpen() and
                not basler_py_driver_node._avena_basler_cameras.IsGrabbing():
                    basler_py_driver_node.get_logger().warn("cameras are not grabbing")
        except KeyboardInterrupt:
            basler_py_driver_node.get_logger().warn("exit key is pressed")
            if basler_py_driver_node._avena_basler_cameras is not None and basler_py_driver_node._avena_basler_cameras.GetSize() > 0:
                basler_py_driver_node.close_basler_cameras(basler_py_driver_node._avena_basler_cameras)
        except ExternalShutdownException:
            basler_py_driver_node.get_logger().warn("external shutdown")
            if basler_py_driver_node._avena_basler_cameras is not None and basler_py_driver_node._avena_basler_cameras.GetSize() > 0:
                basler_py_driver_node.close_basler_cameras(basler_py_driver_node._avena_basler_cameras)
        finally:
            rclpy.try_shutdown()
            basler_py_driver_node.destroy_node()

    """
    def __init__(self, qos_profile:rclpy.qos.QoSProfile,camera_group_serials: dict) -> None:
        super().__init__('basler_camera_controller')
        self.get_logger().info("BaslerPyDriver initialized")
        self._camera_group_serials = camera_group_serials
        self._avena_basler_cameras = None
        self.open_basler_cameras_service_server = self.create_service(Trigger, 'open_basler_cameras', self.open_basler_cameras_cb)
        self.close_basler_cameras_srv = self.create_service(Trigger, 'close_basler_cameras', self.close_basler_cameras_cb)
        self.get_basler_color_image_srv = self.create_service(GetColorImage, 'get_basler_color_image', self.get_basler_color_image_cb)
        self.get_basler_mono_images_srv = self.create_service(GetMonoImages, 'get_basler_mono_images', self.get_basler_mono_images_cb)
        self.get_basler_all_images_srv = self.create_service(GetAllImages, 'get_basler_all_images', self.get_basler_all_images_cb)
        self.basler_camera_color_pub = self.create_publisher(Image, '/basler/color/image_raw', qos_profile)
        self.color_fps = 1.0/37.0 # seconds
        self.timer = None
        self.bridge = CvBridge()
        self._color_image_event_handler = None
        self._left_mono_image_event_handler = None
        self._right_mono_event_handler = None

    def open_basler_cameras(self,camera_group_serials: dict)->pylon.InstantCameraArray:
        """It is responsible for opening multicamera and start grabbing data 
        
        It registers image event handlers for cameras as callback, creates timer
        for publishing color image for security

        Args:
            camera_group_serials: serial numbers for cameras to accquire data
        Returns:
            pylon array of cameras
        Raises:
            GeniCamError: An error occurred accessing busy cameras
        """
        self.get_logger().info("Trying to open cameras")
        self.timer = self.create_timer(self.color_fps, self.basler_color_hd_callback)
        try:
            tlfactory= pylon.TlFactory.GetInstance()
            devices = tlfactory.EnumerateDevices()
            if len(devices) == 0:
                raise pylon.RuntimeException("No camera present.")
            cameras:pylon.InstantCameraArray = pylon.InstantCameraArray(len(devices))
            for i, camera in enumerate(cameras):
                device: pylon.DeviceInfo = tlfactory.CreateDevice(devices[i])
                camera.Attach(device)
                camera_serial = camera.GetDeviceInfo().GetSerialNumber()
                for basler_camera_name , basler_serial_number in camera_group_serials.items():
                    if basler_serial_number == camera_serial and basler_camera_name == 'color':
                        self._color_image_event_handler =ColorImageEventHandler() 
                        camera.RegisterImageEventHandler(self._color_image_event_handler, pylon.RegistrationMode_Append, pylon.Cleanup_Delete)
                    if basler_serial_number == camera_serial and basler_camera_name == 'left_mono':
                        self._left_mono_image_event_handler = LeftMonoImageEventHandler()
                        camera.RegisterImageEventHandler(self._left_mono_image_event_handler, pylon.RegistrationMode_Append, pylon.Cleanup_Delete)
                    if basler_serial_number == camera_serial and basler_camera_name == 'right_mono':
                        self._right_mono_event_handler = RightMonoEventHandler()
                        camera.RegisterImageEventHandler(self._right_mono_event_handler, pylon.RegistrationMode_Append, pylon.Cleanup_Delete)
            cameras.Open()
            for camera in cameras:
                camera = self.configure_basler_camera(camera)
            cameras.StartGrabbing(pylon.GrabStrategy_LatestImageOnly,pylon.GrabLoop_ProvidedByInstantCamera)
            self.get_logger().info("cameras are open")
            return cameras
        except genicam.GenericException as e:
            self.get_logger().error(e.GetDescription())
            self.get_logger().error("cameras failed to open")
            return None

    def open_basler_cameras_cb(self, request: Trigger.Request , response: Trigger.Response) -> Trigger.Response:
        """Callback for opening cameras as a response to service call

        Args:
            request: empty call
            response: done or failed
        Returns:
            succes or failure
        Raises:
            None
        """
        self.get_logger().info("Open Basler cameras is called")
        if self._avena_basler_cameras is None:
            self._avena_basler_cameras = self.open_basler_cameras(self._camera_group_serials)
        else:
            self.get_logger().warn("cameras are already opened, Stop messing around")
        if self._avena_basler_cameras is not None and self._avena_basler_cameras.GetSize() == len(self._camera_group_serials):
            response.success = True
        else:
            response.success = False
        return response

    def close_basler_cameras(self, cameras:pylon.InstantCameraArray) -> None:
        """close cameras and stop grabbing if cameras are open or grabbing

        Args:
            cameras: basler cameras to close 
        Returns:
            None
        Raises:
            None
        """
        if cameras.IsGrabbing():
            cameras.StopGrabbing()
        if cameras.IsOpen():
            cameras.Close()
        self.get_logger().info("cameras are closed")

    def close_basler_cameras_cb(self, request: Trigger.Request , response: Trigger.Response) -> Trigger.Response:
        """Callback for closing cameras as a response to service call

        Args:
            request: empty call
            response: done or failed
        Returns:
            succes or failure
        Raises:
            None
        """
        self.get_logger().info("close Basler cameras is called")
        if self._avena_basler_cameras is not None and self._avena_basler_cameras.GetSize() > 0:
            self.close_basler_cameras(self._avena_basler_cameras)
            self._avena_basler_cameras = None
            self.timer.cancel()
            response.success = True
        else:
            self.get_logger().warn("cameras are already closed or not open")
            response.success = True
        return response


    def get_basler_color_image_cb(self, request:GetColorImage.Request, response: GetColorImage.Response) -> GetColorImage.Response:
        """Callback for geting color image from camera as a response to service call

        Args:
            request: empty call
            response: color image
        Returns:
            color image 
        Raises:
            None
        """
        self.get_logger().info("get_basler_color_image is called")
        if self._color_image_event_handler.color_image is not None:
            response.color = self.bridge.cv2_to_imgmsg(self._color_image_event_handler.color_image, "bgr8") 
        return response

    def get_basler_mono_images_cb(self, request: GetMonoImages.Request, response: GetMonoImages.Response) -> GetMonoImages.Response:
        """Callback for geting mono images from cameras as a response to service call

        Args:
            request: empty call
            response: mono images
        Returns:
            two mono images
        Raises:
            None
        """
        self.get_logger().info("get_basler_mono_images is called")
        if self._left_mono_image_event_handler.left_mono_image is not None:
            response.left_mono =  self.bridge.cv2_to_imgmsg(self._left_mono_image_event_handler.left_mono_image, "mono8") 
        if self._right_mono_event_handler.right_mono_image is not None:
            response.right_mono  =  self.bridge.cv2_to_imgmsg(self._right_mono_event_handler.right_mono_image, "mono8")  
        return response

    def get_basler_all_images_cb(self, request: GetAllImages.Response, response: GetAllImages.Response) -> GetAllImages.Response:
        """Callback for geting all images from cameras as a response to service call

        Args:
            request: empty call
            response: mono and color images
        Returns:
            two mono images and one bgr image
        Raises:
            None
        """
        self.get_logger().info("get_basler_all_images is called")
        if self._color_image_event_handler.color_image is not None:
            response.color = self.bridge.cv2_to_imgmsg(self._color_image_event_handler.color_image, "bgr8") 
        if self._left_mono_image_event_handler.left_mono_image is not None:
            response.left_mono =  self.bridge.cv2_to_imgmsg(self._left_mono_image_event_handler.left_mono_image, "mono8") 
        if self._right_mono_event_handler.right_mono_image is not None:
            response.right_mono  =  self.bridge.cv2_to_imgmsg(self._right_mono_event_handler.right_mono_image, "mono8") 
        return response

    def basler_color_hd_callback(self) -> None:
        """Callback for publishing hd color image for security
        Args:
            None
        Returns:
            None
        Raises:
            None
        """
        if self._color_image_event_handler is not None and self._color_image_event_handler.color_hd_image is not None:
            hd_color_image = self.bridge.cv2_to_imgmsg(self._color_image_event_handler.color_hd_image, "bgr8") 
            now = datetime.now()
            hd_color_image.header.frame_id = "/basler/color"
            hd_color_image.header.stamp.sec =now.second
            hd_color_image.header.stamp.nanosec =now.microsecond//1000
            if self.basler_camera_color_pub.get_subscription_count() > 0 and hd_color_image is not None:
                # Very Slow (average 55 ms) TODO: FIX https://github.com/ros2/rclpy/issues/856
                self.basler_camera_color_pub.publish(hd_color_image)

    def configure_basler_camera(self,camera: pylon.InstantCamera)->pylon.InstantCamera:
        """Configure camera for auto exposure, binning, pixel format and fps
        Args:
            camera: basler camera object
        Returns:
            camera: basler camera object after configuration
        Raises:
            None
        """
        ColorAutoTargetBrightness=0.15
        MonoAutoTargetBrightness=0.2
        MonoBinning=1
        # 18.7 fps x 3840 x 2160 x 1 byte ~= 160000000 bytes/s
        camera.DeviceLinkThroughputLimit.SetValue(320000000) # 37 fps
        # 37f ps x 3840 x 2160 x 1 byte ~= 320000000 byte/s (mono-8)  (3842241000 - 45 fps) # https://docs.baslerweb.com/exposure-time.html
        camera.AutoFunctionROISelector.SetValue("ROI1")
        camera.AutoFunctionROIOffsetX.SetValue(0)
        camera.AutoFunctionROIOffsetY.SetValue(0)
        camera.AutoFunctionROIWidth.SetValue(camera.Width.GetValue())
        camera.AutoFunctionROIHeight.SetValue(camera.Height.GetValue())
        camera.AutoFunctionROIUseBrightness.SetValue(True)
        if camera.ExposureAuto.GetValue() != "Continuous":
            camera.ExposureAuto.SetValue("Continuous")

        if camera.GainAuto.GetValue() != "Continuous":
            camera.GainAuto.SetValue("Continuous")
        
        camera_model = camera.GetDeviceInfo().GetModelName()
        mono_postfix = "um"
        color_postfix = "uc"

        if mono_postfix in camera_model:
            if camera.BinningVertical.GetValue() != MonoBinning:
                camera.BinningVertical.SetValue(MonoBinning)

            if camera.BinningHorizontal.GetValue() != MonoBinning:
                camera.BinningHorizontal.SetValue(MonoBinning)

            if camera.BinningHorizontalMode.GetValue() != "Average":
                camera.BinningHorizontalMode.SetValue("Average")

            if camera.BinningVerticalMode.GetValue() != "Average":
                camera.BinningVerticalMode.SetValue("Average")

            if camera.AutoTargetBrightness.GetValue() != MonoAutoTargetBrightness:
                camera.AutoTargetBrightness.SetValue(MonoAutoTargetBrightness)

        if color_postfix in camera_model:
            if camera.AutoTargetBrightness.GetValue() != ColorAutoTargetBrightness:
                camera.AutoTargetBrightness.SetValue(ColorAutoTargetBrightness)
            camera.PixelFormat = 'BGR8'

        return camera
